crates/typst-kit/src/download.rs

```diff
@@ -0,0 +1,264 @@
+    /// A backlog of the amount of downloaded bytes each second.
+    pub bytes_per_second: VecDeque<usize>,
+    /// The download starting instant.
+    pub start_time: Instant,
+}
+
+/// A minimal https client for downloading various resources.
+pub struct Downloader {
+    user_agent: EcoString,
+    cert_path: Option<PathBuf>,
+    cert: OnceCell<Option<Certificate>>,
+}
+
+impl Debug for Downloader {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        // NOTE: Certificate does not implement Debug, likely for security
+        //       reasons, but we still want to show a field that indicates
+        //       whether a certificate was loaded and is in use, so we implement
+        //       a small helper because `.field_with()` is not yet stable
+        // TODO: remove this once `.field_with` is stabilized (feature =
+        //       debug_closure_helpers)
+        struct CertExists;
+
+        impl Debug for CertExists {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(f, "<Certificate>")
+            }
+        }
+
+        f.debug_struct("Downloader")
+            .field("user_agent", &self.user_agent)
+            .field("cert_path", &self.cert_path)
+            .field("cert", &self.cert.get().and_then(Option::as_ref).map(|_| CertExists))
+            .finish()
+    }
+}
+
+impl Downloader {
+    /// Crates a new downloader with the given user agent and no certificate.
+    pub fn new<S: Into<EcoString>>(user_agent: S) -> Self {
+        Self {
+            user_agent: user_agent.into(),
+            cert_path: None,
+            cert: OnceCell::new(),
+        }
+    }
+
+    /// Crates a new downloader with the given user agent and certificate path.
+    ///
+    /// If the certificate cannot be read it is set to `None`.
+    pub fn with_path<S: Into<EcoString>>(user_agent: S, cert_path: PathBuf) -> Self {
+        Self {
+            user_agent: user_agent.into(),
+            cert_path: Some(cert_path),
+            cert: OnceCell::new(),
+        }
+    }
+
+    /// Crates a new downloader with the given user agent and certificate.
+    pub fn with_cert<S: Into<EcoString>>(user_agent: S, cert: Certificate) -> Self {
+        Self {
+            user_agent: user_agent.into(),
+            cert_path: None,
+            cert: OnceCell::with_value(Some(cert)),
+        }
+    }
+
+    /// Returns the certificate this client is using, if a custom certificate
+    /// is used it is loaded on first access.
+    pub fn cert(&self) -> Option<&Certificate> {
+        self.cert
+            .get_or_init(|| {
+                let pem = std::fs::read(self.cert_path.as_ref()?).ok()?;
+                Certificate::from_pem(&pem).ok()
+            })
+            .as_ref()
+    }
+
+    /// Download binary data from the given url and report its progress.
+    #[allow(clippy::result_large_err)]
+    pub fn download_with_progress(
```

Comment:
I think this should be moved below `download`, basically following the order simple API -> advanced API.